doc/developer/design/20240502_deprecate_subsources.md

```diff
@@ -0,0 +1,190 @@
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR ALL TABLES
+```
+
+`kafka_src` is the progress relation for the source, and a table named `t` is
+automatically created in the same schema as `kafka_src`.
+
+Note that we could also probably make `FOR ALL TABLES` implied for Kafka
+sources.
+
+***
+
+Create a Kafka source that reads from a topic named `t` with a custom table
+name:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR TABLES (t AS my_t)
+```
+
+`kafka_src` is again the progress relation for the source. The data is in a
+table named `my_t`. You would never be able to specify more than one table in
+the original `CREATE SOURCE` statement because you would have to two outputs
+from the same source with the same schema, which we should disallow.
+
+***
+
+Create a PostgreSQL source:
+
+```sql
+CREATE SOURCE postgres_src
+FROM POSTGRES CONNECTION postgres_conn (PUBLICATION p)
+FOR ALL TABLES
+```
+
+`postgres_src` is the progress relation; there is no separate
+`postgres_src_progress` relation created. Otherwise this command works the same
+way that it does today.
+
+***
+
+Add a new table to a PostgreSQL source:
+
+```sql
+CREATE TABLE my_new_table FROM SOURCE postgres_src (upstream_schema.upstream_table);
+```
+
+This would replace the `ALTER SOURCE ... ADD {SUBSOURCE|TABLE}` command.
+
+***
+
+Add a new table to a Kafka source:
+
+```sql
+CREATE TABLE kafka_copy FROM SOURCE kafka_src ("t");
+```
+
+At first this looks weird. Why would you ever want two copies of a Kafka source?
+It turns out this lets you incorporate upstream schema changes without losing
+reclocking information! This would get us reusable reclocking
+(https://github.com/MaterializeInc/materialize/issues/17021), without needing to
+introduce a new `SOURCE PROGRESS` concept.
+
+Note that support for this is blocked on at least
+https://github.com/MaterializeInc/materialize/issues/24843, and also additional
+work to support multi-output sources in Kafka.
+
+#### TEXT + IGNORE COLUMNS
+
+These options can remain on `CREATE SOURCE` statements, and the `ADD SUBSOURCE`
+variants of them will move to instead be `CREATE TABLE` variants. We could also
+probably squeeze #26774 in here without much effort.
+
+### Migration
+
+#### Kafka, single-output load generators
+
+- Current source
+    - ID, persist shard becomes read-only `TABLE`
+    - Durable statement becomes `CREATE TABLE curr_source_name FROM SOURCE
+	curr_progress_name ( topic_name )`
+- Progress collection
+    - ID, persist shard becomes `SOURCE`
+    - Durable statement becomes `CREATE SOURCE curr_progress_name FROM...`
+
+#### PG, MySQL, multi-output load generators
+
+- Current source
+    - ID, persist shard becomes regular `TABLE` no longer associated
+	with the source whatsoever. Users can continue depending on this, but they
+	should drop it after the release at their leisure
+    - Durable statement	becomes `CREATE TABLE curr_source_name ()`
```

Comment:
Do we have any users depending on these sources today? If not ... maybe we can just rip it out?

This seems like a piece of the migration we could do incrementally. E.g., just ship a change that prevents users from depending on sources directly.

Reply:
I don't know if any users depend on them today but this seems like the simplest and most fool-proof approach. I'm not sure what exactly you mean by preventing users from depending on sources directly, though––can you elaborate?